src/libtorch.cc

```diff
@@ -660,7 +672,14 @@ ModelInstanceState::ProcessRequests(
   Execute(&responses, request_count, &input_tensors, &output_tensors);
 
   uint64_t compute_end_ns = 0;
-  SET_TIMESTAMP(compute_end_ns);
+#if defined(TRITON_ENABLE_GPU) && defined(TRITON_ENABLE_STATS)
```

Comment:
Does Execute return before the inference is complete? It doesn't seem so. Is CudaStream() used by pytorch execution engine?

Reply:
Yes.. The Execute does not wait for the inference to complete in case of torch::kGPU is specified. Essentially, its an asynchronous operation. See here for details: https://medium.com/@auro_227/timing-your-pytorch-code-fragments-e1a556e81f2
I couldn't find the exact match solution for libtorch. No documentation regarding `torch::cuda::Event`. 

 > It doesn't seem so. Is CudaStream() used by pytorch execution engine?
 I am confused about that too.. Their documentation says torch uses default stream for the current set device.They define it as the stream where most of the compution occurs. They use CUDAStream abstraction on the streams being used by pytorch execution engine. More details here: https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAStream.h
 
 I did some empirical experiments, and with reported timing values it looks like stream_ callback is deferred till inference run is complete.

Reply:
I guess I can deep -dive into the `c10/cuda/*` files to come with a better solution.

Reply:
I can call [`synchronize` ](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAStream.h#L126) on the CUDAStream object returned by pytorch execution engine to ensure `Execute` remains synchronous. But I think this might impact performance.

Reply:
But we don't have explicit synchronization in the read output function below, unless libtorch does something fancy like implicit synchronization on first output read.

I am surprised that they would use default stream.. I think the launch host func is somehow working because of the CUDA synchronization behavior related to default stream?

Reply:
Yes.. I haven't find it clearly documented, but I believe trying to access `output_tensor` should be internally synchronizing on first output read like : https://github.com/triton-inference-server/pytorch_backend/blob/main/src/libtorch.cc#L867

I have another [commit ](https://github.com/triton-inference-server/pytorch_backend/pull/22/commits/16dc8ef1a45334166b39f91198be0afe7af85437) that ensures the pytorch engine stream is attached the callback..